### PR TITLE
Development

### PR DESCRIPTION
**Pull request by divya-nettem**
Redfish composition service implementation.

### CODE REVIEW

**Review by amar-shalgar — CHANGES_REQUESTED** [review]
Please address comments

File: install/Docker/dockerfiles/scripts/build_odim.sh [hunk 1]
```diff
@@ -17,6 +17,9 @@ LIST=`ls | grep -v 'lib-rest-client' | grep -E '^svc-|^plugin-|add-hosts'`
 echo $LIST
 flag=0
 for i in $LIST; do
+    if [[ "$i" == "svc-composition-service" ]]; then
```
Comment (shivacharanms) [inline install/Docker/dockerfiles/scripts/build_odim.sh:20]: The if condition can be removed as there is no change to 'Dockerfile.odim' and svc-composition-service will not be listed here.

File: install/Docker/dockerfiles/scripts/start_composition_service.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+run_forever()
+{
+        wait
+}
+
+start_composition_service()
+{
+        registry_address="etcd:2379"
+	export PLUGIN_CONFIG_FILE_PATH=/etc/csplugin_config/config.json
+	nohup /bin/svc-composition-service --registry=etcd --registry_address=${registry_address} --server_address=csplugin:45100 --client_request_timeout=`expr $(cat $CONFIG_FILE_PATH | grep SouthBoundRequestTimeoutInSecs | cut -d : -f2 | cut -d , -f1 | tr -d " ")`s  >> /var/log/csplugin_logs/csplugin.log 2>&1 &
```
Comment (shivacharanms) [inline install/Docker/dockerfiles/scripts/start_composition_service.sh:31]: Port 45100 is assigned to a different service, please use the next available port in the preferred series to maintain uniformity. 
Suggested port: 45112

File: install/Docker/dockerfiles/scripts/start_composition_service.sh [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+	PID=$!
+	sleep 3
+
+	nohup /bin/add-hosts -file /tmp/host.append >> /var/log/csplugin_logs/csplugin.log 2>&1 &
```
Comment (shivacharanms) [inline install/Docker/dockerfiles/scripts/start_composition_service.sh:36]: As mentioned above, cs is a service and not a plugin. Please make the changes accordingly.

File: lib-utilities/proto/compositionservice/composition_service.pb.go [hunk 4]
```diff
@@ -0,0 +1,631 @@
+// Code generated by protoc-gen-go. DO NOT EDIT.
```
Comment (shivacharanms) [inline lib-utilities/proto/compositionservice/composition_service.pb.go:1]: Headsup: The *pb.go files will be no longer be part of the github repo as these are generated on the go.

File: lib-utilities/proto/compositionservice/composition_service.proto [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	string URL = 2;
+    bytes RequestBody = 3;
+} 
+
+message CompositionServiceResponse {
+	int32 StatusCode = 1;
+    string StatusMessage = 2;
+    map<string,string> Header=3;
+    bytes Body = 4;  
+}
```
Comment (shivacharanms) [inline lib-utilities/proto/compositionservice/composition_service.proto:47]: Indentation to be handled across the file

File: lib-utilities/services/rpcService.go [hunk 6]
```diff
@@ -364,6 +364,14 @@ func GetEnabledServiceList() map[string]bool {
 			if err == nil && len(resp.Kvs) > 0 {
 				data[microService] = true
 			}
+
+		case "CompositionService":
+			log.Infof("In GetEnabledServiceList Composition service")
```
Comment (shivacharanms) [inline lib-utilities/services/rpcService.go:369]: Log to be removed/ Modified to a better debug log.

File: odim-controller/helmcharts/csplugin/csplugin-config.yaml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+csplugin:
```
Comment (shivacharanms) [inline odim-controller/helmcharts/csplugin/csplugin-config.yaml:1]: Content under helmcharts should be changed to reflect how a service is deployed, looks like these are in line with a plugin. Please compare any odim service and make the changes accordingly.

File: svc-api/models/serviceroot.go [hunk 8]
```diff
@@ -42,6 +42,7 @@ type ServiceRoot struct {
 	OEM                       OEM          `json:"Oem"`
 	RedfishVersion            string       `json:"RedfishVersion"`
 	UUID                      string       `json:"UUID"`
+	CompositionService        *Service     `json:"CompositionService"`
```
Comment (shivacharanms) [inline svc-api/models/serviceroot.go:45]: 'omitempty' needs to added, please refer above lines of code.

**Review by shivacharanms — CHANGES_REQUESTED** [review]
The initial set of comments.

File: svc-composition-service/app/config/config.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
+import os
+import json
+
+CONF_FILE = os.getenv("PLUGIN_CONFIG_FILE_PATH")
```
Comment (shivacharanms) [inline svc-composition-service/app/config/config.py:4]: As mentioned above, please modify your logic to see composition service as a service and not a plugin.
Use service config file, which does not need odim credentials, ideally, service would need certificates for rpc communication and DB details.

File: svc-composition-service/app/config/constants.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
+EVENT_DESTINATION_URL = "/redfish/v1/EventService/CompositionServiceEvent"
```
Comment (shivacharanms) [inline svc-composition-service/app/config/constants.py:1]: Refrain from using 'redfish/v1' for destination URI's

File: svc-composition-service/app/config/constants.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+EVENT_TYPES = [
+    "ResourceRemoved", "ResourceAdded", "ResourceUpdated", "StatusChange",
+    "Alert", "MetricReport", "Other"
+]
+
+EVENT_SUBSCRIPTION_URL = "/redfish/v1/EventService/Subscriptions"
```
Comment (shivacharanms) [inline svc-composition-service/app/config/constants.py:8]: Why declare event service api when an RPC call can be made?

File: svc-composition-service/app/config/constants.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+SYSTEMS_URL = "/redfish/v1/Systems"
```
Comment (shivacharanms) [inline svc-composition-service/app/config/constants.py:10]: Why declare systems service api when an RPC call can be made?

File: svc-composition-service/app/config/constants.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        "MaxCompositions": 1,
+        "NumberOfCompositions": 0
+    },
+    "Client": None,
+    "Pool": "Free"
+}
+
+RESOURCE_ZONE_TEMP = {
+    "@odata.type": "#Zone.v1_6_0.Zone",
+    "Id": "1",
```
Comment (shivacharanms) [inline svc-composition-service/app/config/constants.py:37]: Make sure this value is compliant with redifsh spec.

File: svc-composition-service/app/config/constants.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+EVENT_SUBSCRIPTION_URL = "/redfish/v1/EventService/Subscriptions"
+
+SYSTEMS_URL = "/redfish/v1/Systems"
+
+RESOURCE_BLOCK_TEMP = {
+    "@odata.context": "/redfish/v1/$metadata#ResourceBlock.ResourceBlock.",
+    "@odata.type": "#ResourceBlock.v1_4_0.ResourceBlock",
+    "@odata.id": "/redfish/v1/CompositionService/ResourceBlocks",
+    "Id": "ComputerSystemBlock",
```
Comment (shivacharanms) [inline svc-composition-service/app/config/constants.py:16]: Make sure this value is compliant with redifsh spec.

File: install/Docker/dockerfiles/scripts/start_composition_service.sh (hunk 2 above)
Comment (shivacharanms) [reply install/Docker/dockerfiles/scripts/start_composition_service.sh:31]: Also the composition service should be part of odimra logs, please refer other service start script file to get the path

File: svc-composition-service/app/db/persistant.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+                    port=self.port,
+                    db=PLUGIN_CONFIG["Db"],
+                    socket_timeout=PLUGIN_CONFIG["SocketTimeout"])
+            else:
+                # default host = "localhost", port = 6379
+                self.connection = StrictRedis(
+                    db=PLUGIN_CONFIG["Db"],
+                    socket_timeout=PLUGIN_CONFIG["SocketTimeout"])
+        except Exception as err:
+            logging.error(
```
Comment (shivacharanms) [inline svc-composition-service/app/db/persistant.py:34]: Please make sure the logs are redirected to the log file defined in the start script.

File: svc-composition-service/app/proto/composition_service.proto [hunk 16]
```diff
@@ -0,0 +1,47 @@
+syntax = "proto3";
```
Comment (shivacharanms) [inline svc-composition-service/app/proto/composition_service.proto:1]: Please explain the reasoning behind having the proto file under the composition service. Which is an exact match of the file present under lib-utilities.

File: svc-composition-service/app/rest/resource_zones.py [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+        pipe = self.redis.pipeline()
+
+        if request_data is None:
+            res["Error"] = "Request Body is empty"
+            return res, HTTPStatus.BAD_REQUEST
+
+        try:
+            logging.info("Initialize for creation of Resource Zone")
+
+            # request_data = {
```
Comment (shivacharanms) [inline svc-composition-service/app/rest/resource_zones.py:32]: Remove commented code

**Review by shivacharanms — CHANGES_REQUESTED** [review]
Please address the review comments.

File: install/Docker/dockerfiles/Dockerfile.compositionService [hunk 18]
```diff
@@ -0,0 +1,27 @@
+FROM odim:2.0 as build-stage
+
+FROM cs:2.0 as build-stage-cs
+
+FROM ubuntu:18.04
```
Comment (amar-shalgar) [inline install/Docker/dockerfiles/Dockerfile.compositionService:5]: Please use Ubuntu 20.04

**Review by amar-shalgar — CHANGES_REQUESTED** [review]
Please address comments

**Comment by divya-nettem** [comment]
Raised another pull request. So closing this obsolete pull request.

**Pull request closed without merging**